jetty-servlets/src/main/java/org/eclipse/jetty/servlets/DoSFilter.java

```diff
@@ -1369,10 +1402,77 @@ public void onTimeout(AsyncEvent event) throws IOException
         }
     }
 
-    private String createRemotePortId(ServletRequest request)
+    public enum Action
     {
-        String addr = request.getRemoteAddr();
-        int port = request.getRemotePort();
-        return addr + ":" + port;
+        /**
+         * No action is taken against the Request, it is allowed to be processed normally.
+         */
+        NO_ACTION,
+        /**
+         * The request and response is aborted, no response is sent.
+         */
+        ABORT,
+        /**
+         * The request is rejected by sending an error based on {@link DoSFilter#getTooManyCode()}
+         */
+        REJECT,
+        /**
+         * The request is delayed based on {@link DoSFilter#getDelayMs()}
+         */
+        DELAY,
+        /**
+         * The request is throttled.
+         */
+        THROTTLE;
+
+        /**
+         * Obtain the Action based on configured {@link DoSFilter#getDelayMs()}
+         *
+         * @param delayMs the delay in milliseconds.
+         * @return the Action proposed.
+         */
+        public static Action fromDelay(long delayMs)
+        {
+            if (delayMs < 0)
+                return Action.REJECT;
+
+            if (delayMs == 0)
+                return Action.THROTTLE;
+
+            return Action.DELAY;
+        }
+    }
+
+    /**
+     * Listener for actions taken against specific requests.
+     */
+    public static class Listener
+    {
+        /**
+         * Process the onRequestOverLimit() behavior.
+         *
+         * @param request the request that is over the limit
+         * @param dosFilter the {@link DoSFilter} that this event occurred on
+         * @return the action to actually perform.
+         */
+        public Action onRequestOverLimit(HttpServletRequest request, DoSFilter dosFilter)
```

Comment:
I think it would be good to pass in some info about the actual rate here.
We can't pass in the tracker, as that API is not that great.....
I think passing in the type of the tracker, the number of requests and the period over which there were received may be useful.... even if just to put in the warnings.